Lifetime error complains about a contravariant argument "declared outside the function"

### Code

```Rust
fn weird<'a>(f: impl FnOnce(fn(&'a i32)), x: fn(&'static i32)) {
    f(x);
}
```

### Current output

```Shell
error[E0521]: borrowed data escapes outside of function
 --> src/lib.rs:2:5
  |
1 | fn weird<'a>(f: impl FnOnce(fn(&'a i32)), x: fn(&'static i32)) {
  |          -- lifetime `'a` defined here    - `x` declared here, outside of the function body
2 |     f(x);
  |     ^^^^
  |     |
  |     a temporary borrow escapes the function body here
  |     argument requires that `'a` must outlive `'static`
  |
  = help: `x` is declared outside the function, so any data borrowed inside the function cannot be stored into it

For more information about this error, try `rustc --explain E0521`.
```

### Desired output

```Shell

```

### Rationale and extra context

_No response_

### Other cases

```Rust

```

### Rust Version

```Shell
Reproducible on the playground with version 1.96.0-nightly (2026-03-24 362211dc29abc4e8f8cf)
```

### Anything else?

_No response_

--- a/compiler/rustc_borrowck/src/diagnostics/region_errors.rs
+++ b/compiler/rustc_borrowck/src/diagnostics/region_errors.rs
@@ -686,6 +686,8 @@ impl<'infcx, 'tcx> MirBorrowckCtxt<'_, 'infcx, 'tcx> {
             || (*category == ConstraintCategory::Assignment
                 && self.regioncx.universal_regions().defining_ty.is_fn_def())
             || self.regioncx.universal_regions().defining_ty.is_const()
+            || (fr_name_and_span.is_none()
+                && self.regioncx.universal_regions().defining_ty.is_fn_def())
         {
             return self.report_general_error(errci);
         }

--- a/tests/ui/borrowck/fn-ptr-lifetime-mismatch-with-impl-trait-arg.rs
+++ b/tests/ui/borrowck/fn-ptr-lifetime-mismatch-with-impl-trait-arg.rs
@@ -1,0 +1,8 @@
+// Regression test for https://github.com/rust-lang/rust/issues/154350
+
+fn func<'a>(f: impl FnOnce(fn(&'a i32)), x: fn(&'static i32)) {
+    f(x);
+    //~^ ERROR lifetime may not live long enough
+}
+
+fn main() {}

--- a/tests/ui/borrowck/fn-ptr-lifetime-mismatch-with-impl-trait-arg.stderr
+++ b/tests/ui/borrowck/fn-ptr-lifetime-mismatch-with-impl-trait-arg.stderr
@@ -1,0 +1,10 @@
+error: lifetime may not live long enough
+  --> $DIR/fn-ptr-lifetime-mismatch-with-impl-trait-arg.rs:4:5
+   |
+LL | fn func<'a>(f: impl FnOnce(fn(&'a i32)), x: fn(&'static i32)) {
+   |         -- lifetime `'a` defined here
+LL |     f(x);
+   |     ^^^^ argument requires that `'a` must outlive `'static`
+
+error: aborting due to 1 previous error
+

